In [3]:
# Importar librerías
import pandas as pd

# 1. Analisis inicial de datos

In [4]:
# Cargamos el dataset desde el link
billboard = pd.read_csv('https://raw.githubusercontent.com/hadley/tidy-data/master/data/billboard.csv',  encoding='iso-8859-1')

In [5]:
# Observamos las dimensiones del dataset
billboard.shape

(317, 83)

In [6]:
# Vemos las columnas
billboard.columns

Index(['year', 'artist.inverted', 'track', 'time', 'genre', 'date.entered',
       'date.peaked', 'x1st.week', 'x2nd.week', 'x3rd.week', 'x4th.week',
       'x5th.week', 'x6th.week', 'x7th.week', 'x8th.week', 'x9th.week',
       'x10th.week', 'x11th.week', 'x12th.week', 'x13th.week', 'x14th.week',
       'x15th.week', 'x16th.week', 'x17th.week', 'x18th.week', 'x19th.week',
       'x20th.week', 'x21st.week', 'x22nd.week', 'x23rd.week', 'x24th.week',
       'x25th.week', 'x26th.week', 'x27th.week', 'x28th.week', 'x29th.week',
       'x30th.week', 'x31st.week', 'x32nd.week', 'x33rd.week', 'x34th.week',
       'x35th.week', 'x36th.week', 'x37th.week', 'x38th.week', 'x39th.week',
       'x40th.week', 'x41st.week', 'x42nd.week', 'x43rd.week', 'x44th.week',
       'x45th.week', 'x46th.week', 'x47th.week', 'x48th.week', 'x49th.week',
       'x50th.week', 'x51st.week', 'x52nd.week', 'x53rd.week', 'x54th.week',
       'x55th.week', 'x56th.week', 'x57th.week', 'x58th.week', 'x59th.week',
       '

In [7]:
# Observamos las primeras filas del dataset 
billboard.head()

year      artist.inverted                                  track  time  \
0  2000      Destiny's Child               Independent Women Part I  3:38   
1  2000              Santana                           Maria, Maria  4:18   
2  2000        Savage Garden                     I Knew I Loved You  4:07   
3  2000              Madonna                                  Music  3:45   
4  2000  Aguilera, Christina  Come On Over Baby (All I Want Is You)  3:38   

  genre date.entered date.peaked  x1st.week  x2nd.week  x3rd.week  ...  \
0  Rock   2000-09-23  2000-11-18         78       63.0       49.0  ...   
1  Rock   2000-02-12  2000-04-08         15        8.0        6.0  ...   
2  Rock   1999-10-23  2000-01-29         71       48.0       43.0  ...   
3  Rock   2000-08-12  2000-09-16         41       23.0       18.0  ...   
4  Rock   2000-08-05  2000-10-14         57       47.0       45.0  ...   

   x67th.week  x68th.week  x69th.week  x70th.week  x71st.week  x72nd.week  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

   x73rd.week  x74th.week  x75th.week  x76th.week  
0         NaN         NaN         NaN         NaN  
1         NaN         NaN         NaN         NaN  
2         NaN         NaN         NaN         NaN  
3         NaN         NaN         NaN         NaN  
4         NaN         NaN         NaN         NaN  

[5 rows x 83 columns]

1. **¿Cuáles son las observaciones?**

  Son las canciones que estuvieron el top 100 de billboard 

2. **¿Cuáles son las variables?**

  Año, nombre del artista, canción, duración, género, fecha en la que entró al chart, fecha en la que llegó a la máxima posición, ranking que tuvo en la primera semana,  ranking que tuvo en la segunda semana, etc.



# 2. Preguntas específicas

Para contestar preguntas especificas: realizar filtros y selecciones de columnas existen dos tipos de métodos:

* **Método basado en indexing**: `DataFrame.loc[selección filas, selección columnas]`

* **Método basado en expresiones**: `DataFrame.query(expresión)`

Queremos hacer un análisis sobre estos datos con las siguientes preguntas:

a. ¿Cuáles géneros de música hay?

b. ¿Cuáles son las canciones de género Electronica? 

c. ¿Cuáles son los/las artistas y canciones del género Rock que ingresaron al ranking a partir de octubre? *italicized text*

### a. ¿Cuáles géneros de música hay?

Para contestar esta pregunta podemos usar el método `DataFrame.column.unique()` o la funcion `pd.unique()` que nos permiten obtener los valores únicos de una columna

In [8]:
# Vemos los generos distintos que existen
pd.unique(billboard.genre)

array(['Rock', 'Latin', 'Country', 'Rap', 'Pop', 'Electronica', 'Jazz',
       'R&B', 'Reggae', 'Gospel'], dtype=object)

In [9]:
# La alternativa
billboard.genre.unique()

array(['Rock', 'Latin', 'Country', 'Rap', 'Pop', 'Electronica', 'Jazz',
       'R&B', 'Reggae', 'Gospel'], dtype=object)

### b. ¿Cuáles son las canciones de género Electronica? 
Veamos los dos métodos para contestar esta pregunta

In [10]:
# Metodo basado en indexing
## Ponemos la condicion del filtro
billboard.loc[billboard.genre=='Electronica']

year artist.inverted                  track  time        genre  \
43   2000       Eiffel 65       Blue (Da Ba Dee)  3:29  Electronica   
109  2000    Alice Deejay       Better Off Alone  6:50  Electronica   
265  2000     Fatboy Slim  The Rockafeller Skank  4:00  Electronica   
288  2000  Madison Avenue     Don't Call Me Baby  3:44  Electronica   

    date.entered date.peaked  x1st.week  x2nd.week  x3rd.week  ...  \
43    1999-12-11  2000-01-29         67       29.0       16.0  ...   
109   2000-04-08  2000-06-03         79       65.0       53.0  ...   
265   1999-11-13  2000-01-22         94       94.0       94.0  ...   
288   2000-07-08  2000-09-16         98       96.0       93.0  ...   

     x67th.week  x68th.week  x69th.week  x70th.week  x71st.week  x72nd.week  \
43          NaN         NaN         NaN         NaN         NaN         NaN   
109         NaN         NaN         NaN         NaN         NaN         NaN   
265         NaN         NaN         NaN         NaN         NaN         NaN   
288         NaN         NaN         NaN         NaN         NaN         NaN   

     x73rd.week  x74th.week  x75th.week  x76th.week  
43          NaN         NaN         NaN         NaN  
109         NaN         NaN         NaN         NaN  
265         NaN         NaN         NaN         NaN  
288         NaN         NaN         NaN         NaN  

[4 rows x 83 columns]

In [11]:
# Para seleccionar solo la cancion lo añadimos en el segundo argumento del .loc[]
billboard.loc[billboard.genre=='Electronica', 'track']

43          Blue (Da Ba Dee)
109         Better Off Alone
265    The Rockafeller Skank
288       Don't Call Me Baby
Name: track, dtype: object

In [12]:
# Metodo basado en expresiones
## Definimos el filtro mediante la expresion
billboard.query("genre=='Electronica'")

year artist.inverted                  track  time        genre  \
43   2000       Eiffel 65       Blue (Da Ba Dee)  3:29  Electronica   
109  2000    Alice Deejay       Better Off Alone  6:50  Electronica   
265  2000     Fatboy Slim  The Rockafeller Skank  4:00  Electronica   
288  2000  Madison Avenue     Don't Call Me Baby  3:44  Electronica   

    date.entered date.peaked  x1st.week  x2nd.week  x3rd.week  ...  \
43    1999-12-11  2000-01-29         67       29.0       16.0  ...   
109   2000-04-08  2000-06-03         79       65.0       53.0  ...   
265   1999-11-13  2000-01-22         94       94.0       94.0  ...   
288   2000-07-08  2000-09-16         98       96.0       93.0  ...   

     x67th.week  x68th.week  x69th.week  x70th.week  x71st.week  x72nd.week  \
43          NaN         NaN         NaN         NaN         NaN         NaN   
109         NaN         NaN         NaN         NaN         NaN         NaN   
265         NaN         NaN         NaN         NaN         NaN         NaN   
288         NaN         NaN         NaN         NaN         NaN         NaN   

     x73rd.week  x74th.week  x75th.week  x76th.week  
43          NaN         NaN         NaN         NaN  
109         NaN         NaN         NaN         NaN  
265         NaN         NaN         NaN         NaN  
288         NaN         NaN         NaN         NaN  

[4 rows x 83 columns]

El **método basado en expresiones** puede ser más sencillo de leer pero no se puede seleccionar columnas directamente mientras que el **método basado en indexing** puede ser más complejo de leer pero permite seleccionar filas y columnas directamente

In [13]:
# Luego de utilizar query podemos seleccionar las columnas
billboard.query("genre=='Electronica'")['track']

43          Blue (Da Ba Dee)
109         Better Off Alone
265    The Rockafeller Skank
288       Don't Call Me Baby
Name: track, dtype: object

### c. ¿Cuáles son los/las artistas y canciones del género Rock que ingresaron al ranking a partir de octubre? 

Para contestar esta pregunta debemos filtrar por el genero y la fecha de ingreso y quedarnos con las columnas de artista y canción

Para indicar que se deben cumplir ambas condiciones utilizamos el simbolo `&` entre ambos filtros: (condicion 1) & (condicion 2). Esto significa que se deben cumplir ambas condiciones en simultaneo

In [14]:
# Metodo basado en indexing
billboard.loc[(billboard.genre=='Rock') & (billboard['date.entered'] >= '2000-10-01'), ['artist.inverted','track']]

artist.inverted               track
52   Backstreet Boys, The   Shape Of My Heart
102                 Avant       My First Love
167             Vitamin C            The Itch
170          Thomas, Carl           Emotional
193          3 Doors Down               Loser
211                 Usher       Pop Ya Collar
224           Limp Bizkit             Rollin'
232             Nine Days             If I Am
237        Offspring, The  Original Prankster
247      Wallflowers, The         Sleepwalker
279         Moore, Chante         Straight Up
308         Braxton, Toni      Spanish Guitar
309               Tuesday              I Know

# 3. Preguntas específicas II

Ya pudimos responder nuestra preguntas previas y ahora queremos hacer preguntas más complejas vinculadas a la posición en el ranking en distintas semanas, como por ejemplo:

* ¿En qué semanas estuvo en la posición 1 cada una de las canciones de Rock que ingresaron antes de la segunda mitad del año 2000?

¿Es posible contestar esta pregunta de manera sencilla?

No, parece ser algo bastante complicado.


# 4. Tidy data

Recordemos que las condiciones para que un dataset se considere ordenado (tidy) son:

* Cada variable forma una columna
* Cada observación forma una fila
* Cada tipo de unidad observacional forma una tabla

¿El dataset de billboard cumple con estas condiciones?

No. Observamos que:

1. La variable del número de semana en el ranking no está como una columna
2. Aunque es discutible, podríamos argumentar que hay dos unidades observaciones en la misma tabla: características de la canción y posición en el ranking

Para resolver el primer problema vamos a utilizar la función `pd.melt()`

In [23]:
# Al ejecutar esta celda se abre la documentación de la función
pd.melt?

Signature:
pd.melt(
    frame: 'DataFrame',
    id_vars=None,
    value_vars=None,
    var_name=None,
    value_name: 'Hashable' = 'value',
    col_level=None,
    ignore_index: 'bool' = True,
) -> 'DataFrame'
Docstring:
Unpivot a DataFrame from wide to long format, optionally leaving identifiers set.

This function is useful to massage a DataFrame into a format where one
or more columns are identifier variables (`id_vars`), while all other
columns, considered measured variables (`value_vars`), are "unpivoted" to
the row axis, leaving just two non-identifier columns, 'variable' and
'value'.

Parameters
----------
id_vars : scalar, tuple, list, or ndarray, optional
    Column(s) to use as identifier variables.
value_vars : scalar, tuple, list, or ndarray, optional
    Column(s) to unpivot. If not specified, uses all columns that
    are not set as `id_vars`.
var_name : scalar, default None
    Name to use for the 'variable' column. If None it uses
    ``frame.columns.name`` or 'variable

Nuestro primer paso va a ser especificar las **id_vars**, variables identificadoras y las **value_vars**.

¿Cuáles son cada una de ellas?

**id_vars**: "year", "artist.inverted", "track", "time", "genre", "date.entered", "date.peaked"

Aunque si tenemos en cuenta la pregunta puntual pueden ser menos

**value_vars**: todas las variables de las semanas

Tengamos en cuenta que no es necesario poner la lista de todas las variables ya que la función considera como value_vars a todas aquellas columnas que no aparecen en id_vars

In [16]:
# Aplicamos la función melt y tomamos una muestra
pd.melt(billboard,
        id_vars=["year", "artist.inverted", "track", "time", "genre", "date.entered", "date.peaked"]).sample(5)

year   artist.inverted                  track  time    genre  \
15419  2000       Diffie, Joe  It's Always Somethin'  2:55  Country   
5761   2000      Rimes, LeAnn             I Need You  3:42  Country   
4515   2000          Westlife         Swear It Again  4:07     Rock   
466    2000          Lonestar                  Smile  3:33  Country   
20283  2000  Ghostface Killah       Cherchez LaGhost  3:04      R&B   

      date.entered date.peaked    variable  value  
15419   2000-08-12  2000-09-16  x49th.week    NaN  
5761    2000-05-27  2000-08-12  x19th.week   19.0  
4515    2000-04-01  2000-07-01  x15th.week   20.0  
466     1999-12-18  2000-02-12   x2nd.week   80.0  
20283   2000-08-05  2000-08-05  x64th.week    NaN

Vemos que ahora tenemos el número de la semana como una columna y el ranking en esa semana como otra.

Pero, podemos mejorar el resultado y que nuestras nuevas columnas no tengan nombres genericos como "variable" y "value". Esto lo hacemos utilizando los argumentos **var_name** y **value_name**

In [17]:
# Creamos nuestro dataset tidy
billboard_tidy = pd.melt(billboard, 
                      id_vars=["year", "artist.inverted", "track", "time", "genre", "date.entered", "date.peaked"], 
                      var_name="week", 
                      value_name="ranking")

In [18]:
billboard_tidy.head()

year      artist.inverted                                  track  time  \
0  2000      Destiny's Child               Independent Women Part I  3:38   
1  2000              Santana                           Maria, Maria  4:18   
2  2000        Savage Garden                     I Knew I Loved You  4:07   
3  2000              Madonna                                  Music  3:45   
4  2000  Aguilera, Christina  Come On Over Baby (All I Want Is You)  3:38   

  genre date.entered date.peaked       week  ranking  
0  Rock   2000-09-23  2000-11-18  x1st.week     78.0  
1  Rock   2000-02-12  2000-04-08  x1st.week     15.0  
2  Rock   1999-10-23  2000-01-29  x1st.week     71.0  
3  Rock   2000-08-12  2000-09-16  x1st.week     41.0  
4  Rock   2000-08-05  2000-10-14  x1st.week     57.0

In [19]:
# Dimensiones del dataset
billboard_tidy.shape

(24092, 9)

Ahora el dataset tiene 24092 filas y 9 columnas en comparación a las 317 filas y 83 columnas originales.

Si bien podemos realizar algunas transformaciones adicionales para mejorar el dataset, son necesarias técnicas que veremos más adelante en el curso. Pero ya lo tenemos preparado para contestar la pregunta: ¿En qué semanas estuvo en la posición 1 cada una de las canciones de Rock que ingresaron antes de la segunda mitad del año 2000?

In [20]:
resultado = billboard_tidy.loc[(billboard_tidy['genre']=='Rock')&(billboard_tidy['date.entered']<='2000-06-30')&(billboard_tidy['ranking']==1), ['artist.inverted', 'track', 'week']]
resultado

artist.inverted                  track        week
2227                Sisqo             Incomplete   x8th.week
2230  Aguilera, Christina      What A Girl Wants   x8th.week
2537              Santana           Maria, Maria   x9th.week
2544                Sisqo             Incomplete   x9th.week
2547  Aguilera, Christina      What A Girl Wants   x9th.week
2854              Santana           Maria, Maria  x10th.week
3171              Santana           Maria, Maria  x11th.week
3175                Janet  Doesn't Really Matter  x11th.week
3186        Carey, Mariah  Thank God I Found You  x11th.week
3488              Santana           Maria, Maria  x12th.week
3492                Janet  Doesn't Really Matter  x12th.week
3805              Santana           Maria, Maria  x13th.week
3809                Janet  Doesn't Really Matter  x13th.week
3810      Destiny's Child            Say My Name  x13th.week
3814               N'Sync       It's Gonna Be Me  x13th.week
3819      matchbox twenty                   Bent  x13th.week
4122              Santana           Maria, Maria  x14th.week
4127      Destiny's Child            Say My Name  x14th.week
4131               N'Sync       It's Gonna Be Me  x14th.week
4135              Aaliyah              Try Again  x14th.week
4439              Santana           Maria, Maria  x15th.week
4440        Savage Garden     I Knew I Loved You  x15th.week
4444      Destiny's Child            Say My Name  x15th.week
4756              Santana           Maria, Maria  x16th.week
4757        Savage Garden     I Knew I Loved You  x16th.week
5073              Santana           Maria, Maria  x17th.week
5074        Savage Garden     I Knew I Loved You  x17th.week
5390              Santana           Maria, Maria  x18th.week
5708        Savage Garden     I Knew I Loved You  x19th.week
7937     Vertical Horizon    Everything You Want  x26th.week
8255                Creed    With Arms Wide Open  x27th.week

In [21]:
# Podemos mejorar el resultado ordenando el resultado con la función sort_values
resultado.sort_values(['track', 'week'])

artist.inverted                  track        week
3819      matchbox twenty                   Bent  x13th.week
3175                Janet  Doesn't Really Matter  x11th.week
3492                Janet  Doesn't Really Matter  x12th.week
3809                Janet  Doesn't Really Matter  x13th.week
7937     Vertical Horizon    Everything You Want  x26th.week
4440        Savage Garden     I Knew I Loved You  x15th.week
4757        Savage Garden     I Knew I Loved You  x16th.week
5074        Savage Garden     I Knew I Loved You  x17th.week
5708        Savage Garden     I Knew I Loved You  x19th.week
2227                Sisqo             Incomplete   x8th.week
2544                Sisqo             Incomplete   x9th.week
3814               N'Sync       It's Gonna Be Me  x13th.week
4131               N'Sync       It's Gonna Be Me  x14th.week
2854              Santana           Maria, Maria  x10th.week
3171              Santana           Maria, Maria  x11th.week
3488              Santana           Maria, Maria  x12th.week
3805              Santana           Maria, Maria  x13th.week
4122              Santana           Maria, Maria  x14th.week
4439              Santana           Maria, Maria  x15th.week
4756              Santana           Maria, Maria  x16th.week
5073              Santana           Maria, Maria  x17th.week
5390              Santana           Maria, Maria  x18th.week
2537              Santana           Maria, Maria   x9th.week
3810      Destiny's Child            Say My Name  x13th.week
4127      Destiny's Child            Say My Name  x14th.week
4444      Destiny's Child            Say My Name  x15th.week
3186        Carey, Mariah  Thank God I Found You  x11th.week
4135              Aaliyah              Try Again  x14th.week
2230  Aguilera, Christina      What A Girl Wants   x8th.week
2547  Aguilera, Christina      What A Girl Wants   x9th.week
8255                Creed    With Arms Wide Open  x27th.week